alternative CBOR encoding

yusefnapora opened the pull request:
This is an alternative to #35 that doesn't use the jackson CBOR module.  Instead it uses [cbor-java](https://github.com/c-rack/cbor-java) and provides a `jValueToCbor` method to convert from JValue.  

Doesn't handle the other direction (cbor to json), since that's a lossy conversion.

Commit:
Add hooks

Commit:
Remove sbt-version

vyzo committed:
Merge pull request #32 from mediachain/bigs-cbor

Add simple serialization for datatypes

vyzo committed:
Revert "Add simple serialization for datatypes"

vyzo committed:
Merge pull request #33 from mediachain/revert-32-bigs-cbor

Revert "Add simple serialization for datatypes"

vyzo committed:
Build.scala: add magic incantations for test

vyzo committed:
test scaffolding: BaseSpec, XorMatchers

vyzo committed:
test: JournalIndexSpec

vyzo committed:
Add test spec for journal-commit events

vyzo committed:
test/DummyContext: generate logdirs with mktemp -d

vyzo committed:
tests: better signatures for setup/shutdown functions

vyzo committed:
Merge pull request #34 from mediachain/vyzo-transactor-copycat-tests

Test scaffolding and specs for JournalStateMachine index operations and commit events

yusefnapora commented on transactor/src/main/scala/io/mediachain/util/cbor/JValueConversions.scala:
```diff
@@ -0,0 +1,64 @@
+    throw new MappingException("Unable to convert JValue to CBOR")
+  }
+
+  implicit def jValueToCbor(jValue: JValue): Cbor.DataItem =
```
hmm, this should probably not be implicit...

yusefnapora committed:
remove implicit conversion

yusefnapora committed:
tidy imports

yusefnapora committed:
cleanup JObject conversion

parkan commented on transactor/src/main/scala/io/mediachain/util/cbor/JValueConversions.scala:
```diff
@@ -0,0 +1,62 @@
+      }
+
+      case JDecimal(num) => firstItem {
+        new CborBuilder().add(num.doubleValue())
```
extra parens for non-sideeffecting accessor

yusefnapora committed:
rename firstItem to head, add as implicit method on CborBuilder

yusefnapora committed:
subclass AbstractBuilder so we can call convert directly

vyzo committed:
type support for JournalBlocks

vyzo committed:
StateMachine: Journal Blockchain generation

vyzo committed:
StateMachine: reify Journal block size to a constant

vyzo committed:
JournalBlock: Change entries to Array so that it can be serialized properly

vyzo committed:
StateMachine: Allow configurable blocksize for tests that complete in a reasonable time

vyzo committed:
tests: add Spec for Journal blockchain generation

vyzo committed:
add a quick and dirty test to get measurements of block generation time

yusefnapora committed:
decoder from cbor-java to CValue

yusefnapora committed:
make Reference extend ToCbor trait

vyzo committed:
Merge pull request #40 from mediachain/vyzo-journal-blockchain

Transactor StateMachine: Journal blockchain generation

bigs commented:
oooooooooo i like this! do we trust the lib? if so, go ahead and merge

vyzo commented:
it has conflicts with the changes in types for the blockchain.

yusefnapora committed:
Make ToJObject interface more generic

parkan commented on transactor/src/main/scala/io/mediachain/transactor/Dummies.scala:
```diff
@@ -12,6 +13,10 @@ object Dummies {
     override def hashCode = num
     override def toString = "dummy@" + num
+
+    val CBORType = "dummyReference"
```
this is to be replaced with the real type when it's reserved or?

yusefnapora replied:
I think we'll have to do some more work to get links encoded properly once the tag is reserved.  

[The spec](https://github.com/ipfs/specs/tree/master/ipld#serialised-cbor-with-tags) basically says if there are no properties attached to the `{"@link": "..."}` object apart from the link itself, then you just write out the link value wrapped in the tag.  

But if you do have properties, you write out an array whose first item is the tagged link and whose second item is a map with the remaining properties.

So the complexity sort of depends on whether we want to support properties in the `@link` object.  Either way, we probably don't want to actually write out the `type` field... But this seemed like a decent way to go for now.

vyzo replied:
Our usage in RFC has the references with a naked "@link" without properties.

parkan replied:
I think it's ok to only support naked for now, things still seem not 100% shaken out in IPLD land

yusefnapora committed:
bump json4s version

yusefnapora committed:
use cbor-java to convert from JValue to CBOR

yusefnapora committed:
wrap cbor-java DataItem with CValue type

yusefnapora committed:
add serializers for entries, remove jvalue from types

yusefnapora committed:
add ToCbor trait to all Record types

yusefnapora committed:
pull CValue types into CborAST object

- I suddenly got a bunch of errors about defining case classes, etc
  inside a package object.  I'm not sure why it worked before though...

yusefnapora committed:
just write out `@link` objects without `type` field

yusefnapora committed:
deserialization for DataObjects

yusefnapora committed:
deserialization for journal blocks

yusefnapora committed:
rename ToCbor -> CborSerializable, use as return type of fromCbor methods

yusefnapora committed:
fix copy/paste mistake :)

denisnazarov committed:
Fix link to RFC

parkan committed:
Merge pull request #44 from mediachain/denisnazarov-fixreadme

Fix link to RFC

parkan commented on transactor/src/main/scala/io/mediachain/transactor/StateMachine.scala:
```diff
@@ -12,6 +13,8 @@ import cats.data.Xor
 object StateMachine {
   import io.mediachain.transactor.Types._
+
+  val JournalBlockSize: Int = 4096 // blocksize for Journal Blockchain
```
shouldn't be capitalCCd since it's not a type

vyzo replied:
Let this be here, as it is out of scope for this branch. 
I'll fix it in a separate branch as it has dependent changes in the tests too.

bigs commented on transactor/src/main/scala/io/mediachain/transactor/StateMachine.scala:
```diff
@@ -25,15 +28,23 @@ object StateMachine {
+  case class JournalCurrentBlock() extends Query[JournalBlock]
   
-  case class JournalCommitEvent(entry: JournalEntry) extends Serializable
+  sealed abstract class JournalEvent extends Serializable
```
does this want `CborSerializable` too?

yusefnapora replied:
ah yeah, it probably should be.  I haven't dug into the StateMachine types yet

yusefnapora committed:
add specs2 scalacheck integration, bump specs2 version

bigs commented on transactor/src/main/scala/io/mediachain/transactor/TypeSerialization.scala:
```diff
@@ -0,0 +1,243 @@
+  import io.mediachain.util.cbor.CborAST._
+  import io.mediachain.util.cbor.CborCodec
+
+  object CBORTypeNames {
```
ahhhhhhhhh this is a good spot for this. :+1:

yusefnapora committed:
add spec for round-trip between cbor-java and CValue

- also fix the errors revealed by the spec:
  - remove the CLong type, since there's no equivalent in the CBOR spec
  - match LanguageTaggedString and RationalNumber before CArray
    when converting to CValue, since they're encoded as arrays

yusefnapora commented:
closing to reopen against master
